include/swift/Basic/ExperimentalDependencies.h

```diff
@@ -0,0 +1,968 @@
+  }
+};
+
+/// A general container for double-indexing, used (indirectly) in the
+/// DriverGraph.
+template <typename Key1, typename Key2, typename Value> class TwoStageMap {
+public:
+  // Define this here so it can be changed easily.
+  template <typename Key, typename MapValue>
+  using Map = std::unordered_map<Key, MapValue>;
+
+  using InnerMap = Map<Key2, Value>;
+
+private:
+  Map<Key1, InnerMap> map;
+
+public:
+  Optional<Value> find(const Key1 &k1, const Key2 &k2) const {
+    auto iter = map.find(k1);
+    if (iter == map.end())
+      return None;
+    auto iter2 = iter->second.find(k2);
+    return iter2 == iter->second.end() ? None : Optional<Value>(iter2->second);
+  }
+
+  /// The sought value must be present.
+  Value findAndErase(const Key1 &k1, const Key2 &k2) {
+    auto &submap = map[k1];
+    auto iter = submap.find(k2);
+    assert(iter != submap.end() && "Cannot erase nonexistant element.");
+    Value v = iter->second;
+    submap.erase(iter);
+    return v;
+  }
+
+  bool insert(const Key1 &k1, const Key2 &k2, Value &v) {
+    return map[k1].insert(std::make_pair(k2, v)).second;
+  }
+
+  /// Move value from (old1, old2) to (new1, new2)
+  bool move(const Key1 &old1, const Key2 &old2, const Key1 &new1,
+            const Key2 &new2) {
+    Value v = findAndErase(old1, old2);
+    insert(old1, old2, v);
+  }
+
+  /// Returns the submap at \p k1. May create one if not present.
+  Map<Key2, Value> &operator[](const Key1 &k1) { return map[k1]; }
+
+  /// Invoke \p fn on each Key2 and Value matching (k, *)
+  void forEachValueMatching(
+      const Key1 &k1,
+      function_ref<void(const Key2 &, const Value &)> fn) const {
+    const auto &iter = map.find(k1);
+    if (iter == map.end())
+      return;
+    for (auto &p : iter->second)
+      fn(p.first, p.second);
+  }
+
+  /// Invoke \p fn for each entry
+  void forEachEntry(
+      function_ref<void(const Key1 &, const Key2 &, const Value &)> fn) const {
+    for (const auto &p : map)
+      for (const auto &p2 : p.second)
+        fn(p.first, p2.first, p2.second);
+  }
+
+  /// Invoke fn for each Key1 and submap
+  void
+  forEachKey1(function_ref<void(const Key1 &, const InnerMap &)> fn) const {
+    for (const auto &p : map)
+      fn(p.first, p.second);
+  }
+
+  /// Check integrity and call \p verifyFn for each element, so that element can
+  /// be verified.
+  /// Verify absence of duplicates and call \p verifyFn for each element.
+  void verify(function_ref<void(const Key1 &k1, const Key2 &k2, Value v)>
+                  verifyFn) const {
+    std::unordered_set<Value> vals;
+    for (const auto &p1 : map)
+      for (const auto &p2 : p1.second) {
+        const Key1 &k1 = p1.first;
+        const Key2 &k2 = p2.first;
+        Value const v = p2.second;
+        if (!vals.insert(v).second) {
+          llvm::errs() << "Duplicate value at (" << k1 << ", " << k2 << ") "
+                       << v << "\n";
+          llvm_unreachable("duplicate node");
```

Comment:
This seems like an unusual condition for a general double-keyed map. I feel like it'd be better to just use a `forEachEntry` at the call site.

Reply:
Would you like to clarify? Is "This" supposed to be the double-for loop, or the duplicate detection? Since it is a map, rather than a multimap, the duplication detection seemed relevant here. I could see replacing the double-for loop, but I used it as an independent check on the forEach routine.

Reply:
A multimap allows having multiple entries for the same key, but this is verifying that there are not multiple entries with the same value. If I had a two-keyed map like `TwoStageMap<Type, Type, bool> ConvertibleTypes` that recorded whether the first Type can be converted to the second Type, then of course there are going to be duplicate values.

In your particular use case, you *don't* want duplicate values, but TwoStageMap as a generally-useful data structure doesn't seem like the place to be enforcing that to me.

Reply:
That condition is gone, now. I think it was wrong, as you suspected.